### DISCUSSION TITLE
System.Version input format

### DISCUSSION THREAD

**Discussion by Maximys** [dotnet/runtime#84764] [Q&A]
Hello everybody.
On my PC installed some .NET SDK versions:
`.NET SDKs installed: 
7.0.203 [C:\Program Files\dotnet\sdk]
7.0.300-preview.23179.2 [C:\Program Files\dotnet\sdk]
8.0.100-preview.1.23115.2 [C:\Program Files\dotnet\sdk]`
(part from output _"dotnet --info"_)
and "dotnet --version" returns me:
`8.0.100-preview.1.23115.2`
I expect, that this output should be valid value for [System.Version](https://learn.microsoft.com/en-us/dotnet/api/system.version?view=net-7.0), but when I try to execute:
```
Version version;
version = Version.Parse("8.0.100-preview.1.23115.2");
```
.NET Runtime code throw ArgumentException by [this line](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Version.cs#L314).
Can anybody explain me why it's happend and how I can get .NET version in correct format?

**Comment by Joe4evr** [comment]
> how I can get .NET version in correct format?

![](https://i.imgur.com/p339B41.png)
Not with `System.Version`, that is. `System.Version` only supports a 4-number format, and as you can plainly see, .NET uses a variation of SemVer, which is 3 numbers + optional suffix.

If you want it to be a strong type, you can create your own to parse, or perhaps find a package that defines and parses a SemVer version like this. But really, *why* do you need that?

**Reply by Maximys** [reply]
Just for example, I want to build xunit from source and I run _"./build"_ for than. [Build script](https://github.com/xunit/xunit/blob/main/build) call [build.ps1](https://github.com/xunit/xunit/blob/main/build.ps1) which contain [this line:](https://github.com/xunit/xunit/blob/main/build.ps1#L11):
`$version = [Version]$(& dotnet --version)`
and my build fails with Exception.
Yes, I can temporally hardcode .NET version, but [Version class from Java](https://docs.osgi.org/javadoc/r4v43/core/org/osgi/framework/Version.html) contains _qualifier_ property for _"preview.1.23115.2"_ and I think, .NET\C# should have the same.
[Semantic Versioning 2.0.0](https://semver.org/spec/v2.0.0.html) have next text in Summary:
```
Given a version number MAJOR.MINOR.PATCH, increment the:

MAJOR version when you make incompatible API changes
MINOR version when you add functionality in a backwards compatible manner
PATCH version when you make backwards compatible bug fixes
Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.
```
What do you think?

**Comment by svick** [comment]
Related issues: https://github.com/dotnet/runtime/issues/15868, https://github.com/dotnet/runtime/issues/19317.

**Reply by Maximys** [reply]
Thank you so much for info.
